0000-stream/0000-stream.md

```diff
@@ -0,0 +1,182 @@
+---
+title: STREAM: A Multiplexed Money and Data Transport Protocol
+draft: 1
+---
+
+# STREAM: A Multiplexed Money and Data Transport Protocol
+
+ILP/STREAM, or the STREAMing Transport for the Real-time Exchange of Assets and Messages, is an Interledger transport protocol that provides for multiplexed streams of money and data over a virtual ILP connection. STREAM is a successor to [PSK2](../0025-pre-shared-key-2/0025-pre-shared-key-2.md) and takes significant inspiration from the [QUIC](https://tools.ietf.org/html/draft-ietf-quic-transport-10) Internet transport protocol.
+
+## Definitions
+
+- **Client** - One of the two entities using STREAM. The Client initiates a Connection to the Server using STREAM packets sent over ILP. Because STREAM uses bidirectional streams, the Client may be either the Sender or the Receiver of a given packet.
+- **Connection** - The session established between a Client and a Server. A Connection may have zero or more Money Streams and zero or more Data Streams.
+- **Data Stream** - A bidirectional data channel used for structuring data sent over a Connection. Inspired by [QUIC's Bidirectional Streams](https://tools.ietf.org/html/draft-ietf-quic-transport-10#page-66).
+- **Money Stream** - A bidirectional money "channel" used for sending money over a Connection. Users of STREAM MAY use a single Money Stream for each Connection or multiple to account for different intended purposes for the value sent.
+- **Receiver** - The entity that gets a STREAM packet attached to an ILP Prepare and responds with either an ILP Fulfill or Reject packet. The Receiver may refer to either the Client or Server.
+- **Sender** - The entity that sends a STREAM packet attached to an ILP Prepare. The Sender may refer to either the Client or Server.
+- **Server** - One of the two entities using STREAM. The Server listens through an ILP account for incoming Connections from Clients. Note that STREAM Servers DO NOT listen for Connections over the Internet, but rather over Interledger. Because STREAM uses bidirectional streams, the Server may be either the Sender or the Receiver of a given packet.
+
+## Overview
+
+### Relation to Other Protocols
+
+STREAM is designed to work with ILPv4 Prepare, Fulfill, and Reject packets.
+
+STREAM may be used by applications or higher-level protocols for sending money and data.
+
+### Setup
+
+Before using STREAM, the Client and Server use an authenticated, encrypted communication channel (generally provided as part of an [Application Layer Protocol](../0001-interledger-architecture/0001-interledger-architecture.md#application-layer)) to exchange:
```

Comment:
If you start from a situation where ILPv4 is ubiquitous, then ILPv4 accounting relationships will be everywhere, and it's useful to explain why you would still want to set up an end-to-end piggy-backed protocol on top of a semi-circle ILP path. I discussed this with @BobWay yesterday, and one great application of the semi-circle family-of-protocols (psk, psk2, psk2-draft2, PayStream, and STREAM) is when for instance you scan a barcode (with for instance an ILP address and a public key) on a vending machine: yes, there is an accounting relationship between you and that vending machine, but:

1) the relationship is short-lived
2) there is an air-gapped line-of-sight communication channel in the physical world, and converting that to a data connection (like in https://github.com/interledger/rfcs/pull/186) can be avoided by using a semi-circle payment as a sort of "vpn over ilp"

So then this statement is incorrect; client and server do not need to use a data connection during the setup! :)

As we discussed, you might still use a semi-circle protocol like STREAM to reduce latency, but by saying in the first step that STREAM requires a communication channel, aren't you removing one of its greatest unique selling points?

Reply:
> So then this statement is incorrect; client and server do not need to use a data connection during the setup! :)

Sure, they need a private communication channel where they are sure that no one is eavesdropping. We could add a note that those properties don't need to come from authenticated encryption, but I think in most cases that would just be something like HTTPS (whether the data is sent in the body or headers).

> by saying in the first step that STREAM requires a communication channel, aren't you removing one of its greatest unique selling points?

No, I don't see how it would be. Even if you need to communicate a single message on startup, it makes a big difference to avoid an extra delay for every packet. In many cases that message may not even require an extra roundtrip because it can be sent alongside other application data that you'd be sending anyway. Also, I think we may end up implementing Diffie-Hellman over ILP/STREAM so that you only need to start with a public key and ILP address (that you could get from some kind of public lookup), but I don't think we need to do that right now.

Reply:
Your said elsewhere that "it seems crazy to require a connection over the internet", so then why are you against highlighting that as an advantage here?

Reply:
As I see it the main advantage is that the connection over the internet isn't a part of the protocol flow. Usually you have access to the internet, but because STREAM doesn't need a special purpose channel you can just piggyback the initial setup on top of some communication you were already doing (i.e. in the response headers for an API call, or communicated through a laser)